### ISSUE TITLE
Emails crach the service, maybe missing ssl certificate or wrong mail config?

### ISSUE THREAD

**Issue by rubo77** [freifunk-saar/ff-node-monitor#63]
When trying to add a new node to monitor, I get this error:

```
Error: Response was a non-`Responder` `Err`: Connecting(Io(Custom { kind: Other, error: Ssl(Error { code: ErrorCode(1), cause: Some(Ssl(ErrorStack([Error { code: 337047686, library: "SSL routines", function: "tls_process_server_certificate", reason: "certificate verify failed", file: "../ssl/statem/statem_clnt.c", line: 1915 }]))) }, X509VerifyResult { code: 64, error: "IP address mismatch" }) })).
Warning: Responding with 500 Internal Server Error catcher.

**Comment by rubo77** [comment]
I tried to copy the current https certificate to the VM, but that didn't help:

     scp /etc/letsencrypt/live/freifunk.in-kiel.de/cert.pem 10.77.77.102:/etc/ssl/private/freifunk.in-kiel.de-LE-cert.pem
    update-ca-certificates

**Comment by rubo77** [comment]
They seem to have the same problem here: https://github.com/sfackler/rust-openssl/issues/829

**Comment by RalfJung** [comment]
So this is probably related to sending the email? It sounds like the address given in the cert is not the same as the one that the server actually has. What is your configuration (secrets removed, of course)?

That issue you mention looks different, it doesn't say "IP address mismatch".

**Comment by rubo77** [comment]
In Rocket.toml:

    email_from = "freifunk-knotenalarm@lists.in-kiel.de"

I tried to add a node to my email `freifunk -bei- r.z11.de`

Here is my `$ cat Rocket.toml|grep -v key`:

https://paste.toppoint.de/?ac1784d9d39d4ee8#gfVG0+iDYWgJcvQZFozTQApYSokV7HXB0oJKFRalpSU=

**Comment by RalfJung** [comment]
That is the interesting bit:
```
smtp_host = "10.77.77.101"
```
I have never tried using TLS with an IP address, but I don't think it works very well.
Currently, the only smtp_host for ff-node-monitor that I know will work are "localhost" and hostnames with a valid cert for that hostname. *Maybe* there is a way to get TLS working with bare IP addresses, but I have never tried anything along those lines.

You can do `openssl s_client -connect 10.77.77.101:25 -starttls smtp` to show more information about that certificate.

**Comment by rubo77** [comment]
But this worked before ... So what d might you have changed, that it complains like this now?

**Comment by rubo77** [comment]
Or maybe something changed on our update to Delian buster?

**Comment by RalfJung** [comment]
> this worked before

Oh, I see. Weird.

Yeah, maybe it is related to the update. The `openssl` command I gave above could give a clue to at least see what the error is about.

**Comment by rubo77** [comment]
> `openssl s_client -connect 10.77.77.101:25 -starttls smtp` to show more information about that certificate.

```
CONNECTED(00000003)                                                                                                                                                                                                                 
Can't use SSL_get_servername                                                                                                                                                                                                        
depth=2 O = Digital Signature Trust Co., CN = DST Root CA X3                                                                                                                                                                        
verify return:1                                                                                                                                                                                                                     
depth=1 C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
verify return:1
depth=0 CN = eclabs.de
verify return:1                    
---                        
Certificate chain       
 0 s:CN = eclabs.de                   
   i:C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
 1 s:C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
   i:O = Digital Signature Trust Co., CN = DST Root CA X3
---
Server certificate                                 
-----BEGIN CERTIFICATE-----  
MIIFlzCCBH+gAwIBAgISBPnGVQoJqriynYn2+OFKU3zHMA0GCSqGSIb3DQEBCwUA
...
tlwbiJuW8yYd3456ttWh8gzLCoFgiYuuZLVZ1gPoShtfPedExQX8b42lGYdl9s6Y                                                                                                                                                                    
rk59ZX0vsj/Eeu7nmb7gvZ06IgiYc357Bktns7VcEwGft5XBKl382BvQaw==                                                                                                                                                                        
-----END CERTIFICATE-----                                                                                                                                                                                                           
subject=CN = eclabs.de                                                                                                                                                                                                              
                                                                                                                                                                                                                                    
issuer=C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 3470 bytes and written 442 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE 
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 0838B83416629335236F5AE5EA1691592A1A9550697AD2EEE2181939DE662712
    Session-ID-ctx:
    Master-Key: 59070275BFEEF1488489BA7B71F5B4A91B5EACFC62A71F80CAE82E0BD5D750AB78D4D27822F23AF516D05FB5CBCE0857
    PSK identity: None 
    PSK identity hint: None
    SRP username: None 
    TLS session ticket lifetime hint: 7200 (seconds) 
    TLS session ticket:
    0000 - 0f af 8a 9a 22 63 21 07-99 a0 d5 fd da 33 a2 53   ...."c!......3.S
    0010 - 3d e8 42 9b 59 8d f8 96-d2 d7 ee 46 e7 39 01 42   =.B.Y......F.9.B
...
    0090 - cf 66 7a 34 1b 3d e8 37-6e b4 ef 81 53 2b 98 52   .fz4.=.7n...S+.R
    
    Start Time: 1582545402
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
--- 
250 DSN
```

**Comment by RalfJung** [comment]
> `subject=CN = eclabs.de`

This doesn't look like the certificate is valid for the IP address. It is only valid for the hostname.
And indeed I do not think it is possible to get a cert for an IP address from lets-encrypt.

The certificate has to be valid for whatever address you are giving in the configuration -- that's how SSL hostname validation works.

**Comment by rubo77** [comment]
couldn't we disable SSL on my node monitor? It doesn't make much sense to use SSL locally between two VMs on the same host

**Comment by RalfJung** [comment]
If you find a way to do that with the [`mail` crate](https://docs.rs/mail/0.6.12/mail/), sure. Under the hood this is [`new-tokio-smtp`](https://docs.rs/new-tokio-smtp/0.8.1/new_tokio_smtp) and they are making it quite annoying to create insecure connections.

**Comment by rubo77** [comment]
so where would I have to add those lines? Or do you have to modify the whole node monitor for it?

**Comment by RalfJung** [comment]
The SMTP connection is being set up here:

https://github.com/freifunk-saar/ff-node-monitor/blob/488eddc311c49460e66275143a64adf57f3e5359/src/util.rs#L138-L143

**Comment by rubo77** [comment]
Now I get

     Error: Response was a non-`Responder` `Err`: Connecting(Io(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })).

I tried this hack:

```
// Send email
        let config = if self.config.secrets.get_smtp_host() == "localhost" {
            smtp::ConnectionConfig::builder_local_unencrypted().port(25).build()
        } else {
            // hack by rubo77
            // let smtp_host = self.config.secrets.get_smtp_host();
            // smtp::ConnectionConfig::builder_with_port(smtp_host.parse()?, 25)?.build()
            smtp::ConnectionConfig::builder_local_unencrypted().port(25).build()
        };

**Comment by RalfJung** [comment]
I mean you used `builder_local_unencrypted`. As [the docs for this function](https://docs.rs/new-tokio-smtp/0.8.1/new_tokio_smtp/struct.ConnectionConfig.html#method.builder_local_unencrypted) say

> Creates a connection to 127.0.0.1

Not sure how you expect this to work if you are entirely ignoring `smtp_host`.
You'll have to manually build a [`ConnectionConfig`](https://docs.rs/new-tokio-smtp/0.8.1/new_tokio_smtp/struct.ConnectionConfig.html). All its fields are public so you can instantiate the struct by hand.

**Comment by rubo77** [comment]
we changed the smtp-host to ''mail.eclabs.de" and added the domain to the /etc/hosts file, so it can be resolved, but now we get the error:

```

500: Internal Server Error

The server encountered an internal error while processing this request.
Rocket 
```

and internally if I call `w3m http://0.0.0.0:8833` and add a new node there I get the error

```
422: Unprocessable Entity

The request was well-formed but was unable to be followed due to semantic errors.

━━━━━━━━━━━━━━━━━━━━━━━━━
                                  Rocket

**Comment by RalfJung** [comment]
These errors do not say much, the log that Rocket prints on stdout/stderr should contain some more information.

Also it would help to know which exact URL was requested that lead to these errors.

> `w3m http://0.0.0.0:8833`

`0.0.0.0` is AFAIK not a valid IP address, and even if it is it's most likely not yours, so this looks really suspicious.

**Comment by rubo77** [comment]
This is what Rocket.toml sais:
```
# The address on which ff-node-monitor will listen.  Use "0.0.0.0" to listen on all interfaces.
# Stick with "127.0.0.1" if you are running this behind a reverse proxy (recommended).
address = "0.0.0.0"
```
it works fine if you call 0.0.0.0 via w3m, you can select the node, only after you send the request, you get this error. but I will change it to 127.0.0.1 then

**Comment by rubo77** [comment]
When I set it to 127.0.0.1, I get no response on the website, which is routed via an nginx proxy 

```
upstream knotenalarm {
    server              10.77.77.102:8833;
}
location /knotenalarm/ {
      include                     snippets.d/proxy_generic;
      proxy_pass                  http://knotenalarm/;
}
```
it works fine with 0.0.0.0

**Comment by RalfJung** [comment]
> Use "0.0.0.0" to listen on all interfaces.

Yes, 0.0.0.0 is used on the *listen* side. But I never saw it on the *connecting* side.   And the error "The request was well-formed but was unable to be followed due to semantic errors" sounds to me like it could be caused by a "weird" IP address.

But maybe that guess is wrong.

**Comment by RalfJung** [comment]
> When I set it to 127.0.0.1, I get no response on the website, which is routed via an nginx proxy

Did you change the `Rocket.toml`? No please don't, that's not what I meant! All I said is that using `0.0.0.0` in a *client* is something really odd to do. So keep the `Rocket.toml` as it was but use `w3m http://10.77.77.102:8833` for testing.

**Comment by rubo77** [comment]
OK, I changed it back. But If I use w3m http://10.77.77.102:8833 locally, I also get 

     The request was well-formed but was unable to be followed due to semantic errors.

journalctl sais:

    ....
    Rocket has launched from http://0.0.0.0:8833

I think the URL is not the problem, it is just strange, that you get a different error message from the webserver if you call it locally, than if you call it externally.

**Comment by RalfJung** [comment]
Hm, yes that is strange indeed.

journalctl should usually show more than that when there is an error...

**Comment by RalfJung** [comment]
When I trigger an error on purpose locally, the log looks like this:
```
🚀 Rocket has launched from http://0.0.0.0:8000
Error: Response was a non-`Responder` `Err`: ERROR!!!.
```

**Comment by rubo77** [comment]
I see :

```
Mai 30 10:14:22 freifunk ff-node-monitor[13097]: Error: Response was a non-`Responder` `Err`: Connecting(Io(Custom { kind: Other, error: EsmtpKeyword })).                                                                         
Mai 30 10:14:22 freifunk ff-node-monitor[13097]: Warning: Responding with 500 Internal Server Error catcher.
Mai 30 10:27:04 freifunk ff-node-monitor[13097]: Error: Response was a non-`Responder` `Err`: Connecting(Io(Custom { kind: Other, error: EsmtpKeyword })).                                                                         
Mai 30 10:27:04 freifunk ff-node-monitor[13097]: Warning: Responding with 500 Internal Server Error catcher.
Mai 30 10:27:27 freifunk ff-node-monitor[13097]: Error: Response was a non-`Responder` `Err`: Connecting(Io(Custom { kind: Other, error: EsmtpKeyword })).                                                                         
Mai 30 10:27:27 freifunk ff-node-monitor[13097]: Warning: Responding with 500 Internal Server Error catcher.
Mai 30 10:28:34 freifunk ff-node-monitor[13097]: Error: The incoming form failed to parse.
Mai 30 10:28:34 freifunk ff-node-monitor[13097]: Warning: Responding with 422 Unprocessable Entity catcher.
```
the first was externally triggered when pressing "Hinzufügen" , the last  one was triggered locally at the same event

**Comment by rubo77** [comment]
Maybe because I only allow STARTTLS connections on my mail server?

**Comment by RalfJung** [comment]
Okay so the keyword is `EsmtpKeyword`. The mail client lib I am using doesn't like your mail server.

I have no idea if that lib supports STARTTLS, sorry -- I only ever used it for unencrypted localhost connections.

**Comment by RalfJung** [comment]
Looks like the error is raised when it finds what is considers to be "invalid characters" in the ESMTP announcement, or so. Do you have anything odd in the HELO hostname that your mail server announces?

**Comment by rubo77** [comment]
what would be the command to elaborat this?

**Comment by RalfJung** [comment]
Mostly your mail server config / server hostname. I think there is a way to use `openssl` to manually become a client of an SMTP server and speak SMTP yourself, but I don't have the time to write detailed instructions for that, sorry.

**Comment by rubo77** [comment]
telnet works:

```
$ telnet ****.de 25
Trying 2a01:4f8:****:254...
Trying 116.****...
Connected to ****.de.
Escape character is '^]'.
220 ****.de ESMTP Postfix (Debian/GNU)
HELO
501 Syntax: HELO hostname
HELO ****.de
250 ****.de
MAIL FROM: <poptest@****.de>
250 2.1.0 Ok
RCPT TO: <ruben.barkow@****.de>
250 2.1.5 Ok
DATA
354 End data with <CR><LF>.<CR><LF>
test
.
250 2.0.0 Ok: queued as 520C97E05F2
```

**Comment by RalfJung** [comment]
```
220 ****.de ESMTP Postfix (Debian/GNU)
```
Something about that line, I think, is what it does not like. But I might also be guessing wrong. You don't happen to have an umlaut in the hostname or so?

**Comment by RalfJung** [comment]
Hm no that can't be it, as it would reject even `.`. So this is validating something else.

It is probably a bug in https://github.com/dac-gmbh/new-tokio-smtp...

**Comment by rubo77** [comment]
I found out, I still had this local changes in my `src/util.rs`:

```
-            let smtp_host = self.config.secrets.get_smtp_host();
-            smtp::ConnectionConfig::builder_with_port(smtp_host.parse()?, 25)?.build()
+            smtp::ConnectionConfig::builder_with_port(self.config.secrets.get_smtp_host().parse()?, 25)?.build()
```

i reverted that but a service restart didn't change anything, now I pulled the latest master and will rebuild now...

**Comment by rubo77** [comment]
to be exact, the error message format changed a bit, but still the same overall message:

```
Error: Response was a non-`Responder` `Err`: I/O-Error: syntax error parsing esmtp-keyword from str.
Warning: This `Responder` implementation has been deprecated.
Warning: In Rocket v0.5, `Result<T, E>` implements `Responder` only if `E` implements `Responder`. For the previous behavior, use `Result<T, Debug<E>>` where `Debug` is `rocket::response::Debug`.
Warning: Responding with 500 Internal Server Error catcher.
```

**Comment by rubo77** [comment]
couldn't you add more output somehow to the error message? so we see the string in the error message that is causing the error?

**Comment by RalfJung** [comment]
I can't print anything that the library emitting the error doesn't tell me. And the library only says `Connecting(Io(Custom { kind: Other, error: EsmtpKeyword }))`.

With the newer version, as you can see this has gotten prettier:
```
I/O-Error: syntax error parsing esmtp-keyword from str.
```
But I cannot show any information that is not provided by the underlying library.

**Comment by RalfJung** [comment]
Since this email library doesn't seem to be great, I do want to move away from it. I plan to switch back to https://github.com/lettre/lettre. However, the reason I moved away from that is that it had trouble with non-ASCII characters.  That trouble is fixed by https://github.com/lettre/lettre/pull/383, but so far no release with the fix exists.

One lettre 0.10 is released, I'll port ff-node-monitor to that, and hopefully that should help. But naturally I have no idea when that will be the case.

**Comment by rustonaut** [comment]
Note: That new-tokio-smtp, is _not_ a email library, it only does SMTP and nothing more and is meant to be used mainly by more general email libraries. :wink:  

So if you are using it directly as a mail library it's probably not the best idea as it doesn't provide any way to format/create a mail body.

**Comment by rustonaut** [comment]
Just to be clear I'm still looking into the bug anyway.

It seems to be a problem with EHLO parsing.

Note that the telnet log above only used HELO so it's not showing the part the library is choking on.

I also plan to make parsing failures there less fatal and log them.

**Comment by RalfJung** [comment]
> So if you are using it directly as a mail library it's probably not the best idea as it doesn't provide any way to format/create a mail body.

I'm not, I just tracked this problem down to there.
I'm using your mail crate: https://crates.io/crates/mail

**Comment by rubo77** [comment]
The reacted information ***** only contained ascii letters

**Comment by rubo77** [comment]
The telnet output with EHLO is the same as above, onyl another start:
```
EHLO eclabs.de
220 eclabs.de ESMTP Postfix (Debian/GNU)
250-eclabs.de
250-PIPELINING
250-SIZE 90000000
250-VRFY
250-ETRN
250-STARTTLS
250-ENHANCEDSTATUSCODES
250-8BITMIME
250 DSN
MAIL FROM: <poptest@...
```

**Comment by rustonaut** [comment]
Can I assume that `220 ****.de ESMTP Postfix (Debian/GNU)` is actually the initial server greeting, which sneaked below the EHLO line because you send it faster then then initial server response?

(Given the error cough it should be, I think)

**Comment by rubo77** [comment]
This is directly copied from my telnet console

**Comment by RalfJung** [comment]
@rubo77  I have updated the new-tokio-smtp dependency of ff-node-monitor, so you could try again with the latest version. (But I am not sure if logging will work, ff-node-monitor doesn't set up the log crate.)

**Comment by rubo77** [comment]
GEIL geht!!!!

**Comment by rubo77** [comment]
Now I get a similar error when trying to add monitor in the vagrant mashine:

```
Error: Response was a non-`Responder` `Err`: I/O-Error: Connection refused (os error 111).
Warning: This `Responder` implementation has been deprecated.
Warning: In Rocket v0.5, `Result<T, E>` implements `Responder` only if `E` implements `Responder`. For the previous behavior, use `Result<T, Debug<E>>` where `Debug` is `rocket::response::Debug`.
Warning: Responding with 500 Internal Server Error catcher.
```

I defined the example configuration for msmtp from the Readme and used my googlemail account, which has  2FA  disabled

**Comment by RalfJung** [comment]
I doubt this is always the same issue, so please open a new issue for new problems.

This time, the error is:
```
I/O-Error: Connection refused (os error 111)
```
So, something about your system configuration seems to be wrong.

**Comment by rubo77** [comment]
just for info:

up until yesterday, there was sill this error warning in the journal log:

    Warning: Parsing Server EHLO response partially failed: syntax error parsing esmtp-keyword from str

since the last update this is gone too and everything works

**Comment by RalfJung** [comment]
Ah, that is the warning I meant in https://github.com/dac-gmbh/new-tokio-smtp/issues/5#issuecomment-643774147. Looks like log doesn't even need setup to print something, nice.

But the latest update only changed log settings, that shouldn't make the difference here...

**Comment by RalfJung** [comment]
>  Any log events that occur before initialization will be ignored.

Hm no, you really shouldn't be getting any warnings before the update... or does the lib also print directly to stdout or so, @rustonaut  ?
